test/test_dynamic_transform.cpp

```diff
@@ -618,4 +618,217 @@ TEST_F(NVFuserTest, DynamicTransform11_CUDA) {
+    fusion.addInput(output_shape[i]);
+  }
+  auto x_reshape = reshape(tv1, output_shape);
+  auto y = (reshape_before_reduction) ? sum(x_reshape, {kReductionAxis})
+                                      : add(x_reshape, bias);
+  fusion.addOutput(y);
+
+  FusionExecutorCache fusion_executor_cache(std::move(fusion_ptr));
+
+  // Return pair of: number of concretizations & total number of kernel runtimes
+  auto countConcretizations = [&fusion_executor_cache]() {
+    std::unordered_set<const std::pair<
+        size_t,
+        std::optional<DynamicTransformConcretizationInfo>>*>
+        concs;
+    for (auto& it : fusion_executor_cache.getKernelRuntimes()) {
+      concs.insert(&it.first);
+    }
+    return concs.size();
+  };
+  size_t num_concretizations = countConcretizations();
+  // Check that concretizations and runtimes are cache misses only when they
+  // should be
+  auto checkCache = [&countConcretizations,
+                     &num_concretizations](bool expect_miss) {
+    auto current = countConcretizations();
+    ASSERT_EQ(current, num_concretizations + (size_t)expect_miss);
+    num_concretizations = current;
+  };
+
+  for (auto& inv : invocations) {
+    auto input_shape = std::get<0>(inv);
+    auto output_shape = std::get<1>(inv);
+    auto expect_miss = std::get<2>(inv);
+
+    auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+    auto bias_shape = (reshape_before_reduction) ? input_shape : output_shape;
+    at::Tensor at_x = at::randn(input_shape, options);
+    at::Tensor at_bias = at::randn(bias_shape, options);
+    std::vector<c10::IValue> aten_inputs = {at_x, at_bias};
+    // Add input scalars describing the reshape size for concretization
+    for (int i : c10::irange(output_dims)) {
+      aten_inputs.push_back(output_shape[i]);
+    }
+
+    auto outputs = fusion_executor_cache.runFusionWithInputs(aten_inputs);
+    checkCache(expect_miss);
+
+    auto at_tv1 = (reshape_before_reduction) ? (at_x + at_bias)
+                                             : at::sum(at_x, kReductionAxis);
+    auto at_x_reshape = at::native::view(at_tv1, output_shape);
+    auto at_y = (reshape_before_reduction)
+        ? at::sum(at_x_reshape, kReductionAxis)
+        : at::add(at_x_reshape, at_bias);
+
+    testValidate(&fusion, outputs, aten_inputs, {at_y}, __LINE__, __FILE__);
+  }
+}
+
+TEST_F(NVFuserTest, FusionDynamicReshapeReductionShmoo_CUDA) {
+  std::vector<std::pair<std::vector<int64_t>, std::vector<int64_t>>>
+      all_dynamic_reshape_examples = {};
+  for (auto invocations : std::vector<std::vector<dynamic_view_invocation>>{{
+           {{8, 3 * 4, 7, 9}, {8, 3 * 4, 7, 9}, true}, // trivial
+           {{8, 3 * 4, 7, 5}, {8, 3 * 4, 7, 5}, false}, // trivial
+           {{8, 3 * 4, 7, 9}, {8, 3, 4, 7 * 9}, true}, // merge(2) osplit(1, 3)
+           {{8, 3 * 4, 7, 9},
+            {8, 3, 4 * 7, 9},
+            true}, // merge(1) merge(2) osplit(1, 3)
```

Comment:
Added this shmoo test. It revealed a bug where this particular case failed with an unexpected cache hit. In this case the `cacheId` of `KernelArgumentHolder` was always computed without input scalars and caused an early short-circuit, but we are sending the output shape as an input scalar here, so the concretization depends on input scalars. In case of dynamic fusions, we now add int scalar inputs to the `cacheId` string.

Reply:
Thanks. We could analyze the fusion so that only int parameters that could matter be used for the caching, but don't know if it's important. Thanks for the fix.

Reply:
Yes, we use all int scalar inputs now, which means that if one of those changes but it's not involved in a reshape, it will trigger a rebuild of concretization info. There may be cases where this happens; I had in mind that most scalars would not change iteration to iteration. I'll have a look at what that would take.

Reply:
I think this is not something urgent for now. The latency of `getConcretizationInfo` would be more important to address, for example. Please just leave a comment about why it's not ideal for now.